compiler/rustc_const_eval/src/interpret/cast.rs

```diff
@@ -133,6 +133,25 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                     bug!()
                 }
             }
+
+            Transmute => {
+                assert!(src.layout.is_sized());
+                assert!(dest.layout.is_sized());
+                if src.layout.size != dest.layout.size {
+                    throw_ub_format!(
+                        "transmuting from {}-byte type to {}-byte type: `{}` -> `{}`",
+                        src.layout.size.bytes(),
+                        dest.layout.size.bytes(),
+                        src.layout.ty,
+                        dest.layout.ty,
+                    );
+                }
+                if dest.layout.abi.is_uninhabited() {
+                    throw_ub_format!("transmuting to uninhabited type `{}`", dest.layout.ty,);
+                }
```

Comment:
This is a somewhat strange special case. Previously we *had to* have such an error since we simply didn't have a `ret` block to jump to; now this leads to some MIR validation being done even with validation disabled (that's the test case behavior change you mentioned), which seems rather surprising.

Reply:
Not sure which part you mean should be done in a separate PR, but this here is responsible for the change in when [never_transmute_void.rs](https://github.com/rust-lang/rust/pull/108442/files#diff-ade4cecd2b4aa3e7a9add16038ec867852852f808d755d317365bc27f58d0d8a) reports an error, and I don't think that's a good change.

Reply:
The problem I was hitting here is that this is the only place I could come up with to put the "transmuting to uninhabited type" message that was previously there from

https://github.com/rust-lang/rust/blob/13b7aa4d7faafbed78d195c27f6868f6bd90c2f8/compiler/rustc_const_eval/src/interpret/intrinsics.rs#L130

And thus keep the nice message from

https://github.com/rust-lang/rust/blob/13b7aa4d7faafbed78d195c27f6868f6bd90c2f8/tests/ui/consts/const-eval/ub-enum.32bit.stderr#L110-L111

I guess the thing that's not clicking with me is why it's not considered validation for MIR building to take away call targets for uninhabited destination places -- the transmute *could* have had a target block the whole time -- but it would be considered validation to look at the *type* layout information here.

If this was breaking `-Zmiri-disable-validation` for something like `-1_i32 as char (Transmute)`, then I'd totally agree that it would be wrong.  But this check is still entirely *type* based, like it done elsewhere in MIR, and is never looking at a *value* to validate it.

(And from an implementation perspective, this is just looking at the layout information that it *must* calculate anyway in order to interpret the transmute.  If it needed an extra `inhabited_predicate` query call to go try to figure it out, that would be extra evidence that this was doing spurious checking that it shouldn't.  But to me it's basically the same as the "same sizes" check: conceptually an uninhabited type is -∞ bytes, so transmuting to an uninhabited type is UB for not matching the size of the input, regardless of the values involved.)

Might it be ok to leave the implementation this way, but where I add new miri tests that transmute to `&Void` instead to demonstrate that's disallowed by default but ok under `-Zmiri-disable-validation`?

Reply:
The problem I was hitting here is that this is the only place I could come up with to put the "transmuting to uninhabited type" message that was previously there from

https://github.com/rust-lang/rust/blob/13b7aa4d7faafbed78d195c27f6868f6bd90c2f8/compiler/rustc_const_eval/src/interpret/intrinsics.rs#L130

And thus keep the nice message from

https://github.com/rust-lang/rust/blob/13b7aa4d7faafbed78d195c27f6868f6bd90c2f8/tests/ui/consts/const-eval/ub-enum.32bit.stderr#L110-L111

I guess the thing that's not clicking with me is why it's not considered validation for MIR building to take away call targets for uninhabited destination places -- the transmute *could* have had a target block the whole time -- but it would be considered validation to look at the *type* layout information here.

If this was breaking `-Zmiri-disable-validation` for something like `-1_i32 as char (Transmute)`, then I'd totally agree that it would be wrong.  But this check is still entirely *type* based, like it done elsewhere in MIR, and is never looking at a *value* to validate it.

(And from an implementation perspective, this is just looking at the layout information that it *must* calculate anyway in order to interpret the transmute.  If it needed an extra `inhabited_predicate` query call to go try to figure it out, that would be extra evidence that this was doing spurious checking that it shouldn't.  But to me it's basically the same as the "same sizes" check: conceptually an uninhabited type is -∞ bytes, so transmuting to an uninhabited type is UB for not matching the size of the input, regardless of the values involved.)

Might it be ok to leave the implementation this way, but where I add new miri tests that transmute to `&Void` instead to demonstrate that's disallowed by default but ok under `-Zmiri-disable-validation`?

Reply:
> I guess the thing that's not clicking with me is why it's not considered validation for MIR building to take away call targets for uninhabited destination places -- the transmute could have had a target block the whole time -- but it would be considered validation to look at the type layout information here.

Because if there is no call target we don't have any other choice. :shrug: Also it's completely different checks that have nothing to do with each other in Miri: one is traversing the value to see if it matches the type, one is simply figuring out what to do next after an intrinsic returned.

So, I never considered this a "nice message". It was a necessity to hack around the fact that `transmute` sometimes lacked a target block.

> the transmute could have had a target block the whole time

That's my point exactly -- if it did, we would never have added this message. I'd rather not go out of my way now to preserve this historical accident.

>  conceptually an uninhabited type is -∞ bytes

That model fails very quickly in Rust, e.g. `(!, u8)` has size 1 and it is crucial for soundness that it does (due to partial initialization). So I don't think this is a helpful guide.

There are tons of places all over Miri where we could add extra checks of the form `if !inhabited { UB }`; I think it is almost always a mistake to do that (and I have removed some of these checks). Uninhabited types are not special; they should only raise UB by virtue of their validity check never passing -- this is conceptually the same UB that also triggers on, e.g., transmuting 2u8 to `bool`. Any time we have to special case uninhabited types to me indicates we are doing something wrong and not properly treating all types uniformly, instead adding special ad-hoc hacks. I don't think that's ever a good thing. This `if` you added here is an example of exactly that.